### DISCUSSION TITLE
Navigation distance settings

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#108] [Q&A] [answered]
_Cruiser is about user choice_: allow users to adjust the distances of voice navigation according to their needs.

The process is now much simpler:
- Turn on / off the navigation instructions you want
- Adjust the distances of the navigation instructions
(separately by navigation profile)

See the default distances (car is similar to motorcycle):

<img src="https://user-images.githubusercontent.com/3484020/228798416-ad310311-c927-4089-8895-a081d4c89b87.jpg" width="33%"/><img src="https://user-images.githubusercontent.com/3484020/228798427-54f20a03-3a9c-41aa-8727-402b19ec9f24.jpg" width="33%"/><img src="https://user-images.githubusercontent.com/3484020/228798435-ce42f748-ebfa-4124-be7c-791f9f82bcbd.jpg" width="33%"/>
<img src="https://user-images.githubusercontent.com/3484020/228832815-854a48a7-a753-4529-98f0-cfdc3fca2d01.jpg" width="33%"/><img src="https://user-images.githubusercontent.com/3484020/231707687-57c79dff-ccce-4c84-b6f0-ef743af8ff55.jpg" width="33%"/><img src="https://user-images.githubusercontent.com/3484020/231707702-a1ee914b-3964-43fb-90ab-490766033f9d.jpg" width="33%"/>

**Comment by devemux86** [comment]
With these new settings, everything should be simpler:
- Average speed of profile only affects the duration of the route
(enable it in "Settings | Routing | Options | Route duration")
- Announcement time (early - late) setting is no longer needed

**Reply by WolfgangZ42** [reply]
Although I was fine with the current procedure - a simplification. Perfect!

**Comment by devemux86** [comment]
Please see the default distances (car is similar to motorcycle) and let me know if we need to change anything.

Or simplification: e.g. "Turn now" and "Arrive at" distances are very close,
we could use the "Turn now" distance for "Arrive at" or keep them separate?

**Reply by WolfgangZ42** [reply]
> use the "Turn now" distance for "Arrive at"

should be sufficient as "last announcement" (but that sounds quite final 😁)

**Reply by devemux86** [reply]
> should be sufficient as "last announcement"

Nice idea, "Last announcement" and "Arrival announcement" are better descriptions, I updated the images in:

- https://github.com/devemux86/cruiser/discussions/108#discussion-5022326

**Comment by 0709wiwiwi** [comment]
I suppose Arrive is at the destination >  only anounces the END of the track ? 
![image](https://user-images.githubusercontent.com/54167201/228835286-22840282-3044-4604-9a21-aa698f829b6c.png)
I compare with other app arriving at destination:  1. By the small distance  (like turns) + arrive at destination. (Single default no menu)

The Turn now  0 = none.  Possibility to just set single announcement before turns.  As the best TTS = Less TTS. 
At slow (walk) speeds a too small distance before a turn can have the effect that the announcement arrives after the turn. (gps tolerances).
@devemux86 
**Additional question.** 
Does the last minute avg speed has a correcting  effect on the announcement distances or  timing ?

**Reply by devemux86** [reply]
> I suppose Arrive is at the destination > only anncounces the END of the track ?

Correctly.
We could use this distance for via points too, now they use the "Last announcement".

> Does the last minute avg speed has a correcting effect on the announcement distances or timing ?

- **Average** speed (selected by user) does not affect the voice announcements anymore.
**Average** speed only changes the duration of calculated or imported routes (if enabled).
See: https://github.com/devemux86/cruiser/discussions/108#discussioncomment-5475943

- Voice distances can now be selected precisely by users (instead of automagically calculated)

- **Current** speed (if enabled) allows the speed from GPS to influence the voice guidance:
speak the voice announcements some seconds before the position has been reached.
(Settings | Voice guidance | Advanced | Current speed)

**Reply by 0709wiwiwi** [reply]
> We could use this distance for via points too, now they use the "Last announcement".

Simple: Announce Via points 100% identical just like announcing the turns.

**Reply by devemux86** [reply]
> Simple: Announce Via points 100% identical just like announcing the turns.

This is how it works right now and continues to work in the above new distance settings. :slightly_smiling_face:

**Reply by 0709wiwiwi** [reply]
> Does the last minute avg speed has a correcting effect on the announcement distances or timing ?

Sorry I had to ask for ;-). 
Comment.
As you probably know Locus has a genial variable and adaptable distance announcement (timing in seconds) that is triggered by the momentary (last minute) avg speed. 
The disadvantages ? 
This is more an concern so for cyclist and hikers.
At slow speed walking the announcement sometimes is too late (gps vs map tolerances)  
For cyclist who after a climb go downhill at lightning fast speeds also the announcement can come way too late (for a minute or so)

**Reply by 0709wiwiwi** [reply]
Info.
https://forum.kurviger.de/t/kurviger-app-3-0-6-feedback/10720/101?

> **Additional question.**
> Does the last minute avg speed has a correcting effect on the announcement distances or timing ?

That was what I was thinking about and why I so supposed that the following feature was already present into Cruiser ;-)
Settings &gt; Voice guidance &gt; Advanced &gt; Current speed &gt; "Influence of speed on voice guidance" &gt; (Un)Set

**Reply by devemux86** [reply]
I wrote the first implementation of Cruiser navigation many years ago,
before the appearance of the app on Google Play in 2013.

Since then it has been continuously improving based on user needs.

Now all options are in the hands of the users with detailed settings:
- Voice distances
- Influence of speed
- Speed auto zoom
- ...

**Reply by WolfgangZ42** [reply]
> Does the last minute avg speed has a correcting effect on the announcement distances or timing ?

And 50 seconds of the last minute staying on a red light???
In my opinion we should stick to the distance in meter - I use 1500 300 50 ... and I know how fast I am.

**Reply by 0709wiwiwi** [reply]
You are very correct Wolfgang.  As I know the disadvantages observed in the other app ;-)
I expect. 
1. Optional. In Advanced - Current Speed:  Unset or Set.
2. The dynamic change range must be narrowed to the lowest speed as is  set by the user. Red light ? = Virtualy continue ;-)
By Cruiser speed setting:  "**min** avg max".  The reference avg speed used for THIS calculation should not go down to zero but hangs on the **min** set.
3. To keep the method stable.  Influences just the TTS announcement timing ! (Not changing the prefered distances commands spoken). At every (high) speed the distance announcement should be completed just before reaching the command distances set. (Sorry is so hard to describe I hope you understand ?) Or the TTS message sequence has to be triggered earlier or later. 
As some languages (DE ?) expect more words than others ? No ? 
As I am a slow speed cyclist, only you the Motorbikers can confirm the current solution  is fine or not ? 
On slow cycling speed I notice no change so....pse bikers...you test)

_Turn commands long or short ? 
My car gps (DE). The turn command are usually followed by friendly NL "alstublieft".  (Please)
The TTS lady drives me crazy with her kindness.  Who did program something like this ?_

**Reply by devemux86** [reply]
The previous voice announcement system (described [here](https://github.com/devemux86/cruiser/discussions/19)) used dynamic distances based on _current_ speed.
(can still be available as an option!)

Current voice announcement system (described [above](https://github.com/devemux86/cruiser/discussions/108)) allows users to adjust everything:
- Voice distances
- Influence of speed

>  Influences just the TTS announcement timing !

Precisely, _current_ speed setting (if enabled) can speak instructions earlier if driving fast.

**Comment by devemux86** [comment, marked as answer]
Navigation distance settings are implemented in [Cruiser 3.0.16](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

"Settings | Voice guidance | Distances" or search settings for "distance".

---

With the new voice distance settings and the new customized [speed auto zoom](https://github.com/devemux86/cruiser/discussions/63) (normal or [reverse](https://github.com/devemux86/cruiser/discussions/88)),
you are now in full control of navigation with Cruiser according to your driving style and preferences!

**Reply by DolomitesTom** [reply]
need more testing, but looks good

**Reply by devemux86** [reply]
Indeed, please test it well, as it affects the core mechanism of voice navigation.

**Reply by 0709wiwiwi** [reply]
Examples by set Bike.
Voice guidance (TTS)
A single TTS announcement/turn is now possible.  :-))
- By Last announcement + Navigation instructions set off.  **Off ?** 

**See the video.**  
https://youtube.com/shorts/rC3rgoAm1og
_The second (next) turn announcement is a little bit too early .
With a Left/Right combination, if the Right comes too early it would be confusing
The next turn announcement should happen AFTER passing the first left turn._

**Navigation instructions off ?  = False**
_As there remain 2 announcements:  
The Last announcement AND the Arrival announcement._

**Proposal.** 
Simplify and unify the settings into a single panel. 
Into distances: By varinants > Car- Motorbike - Bike - Walk.
**Advantage**: 
By select transport: > Optimal setting/transport is automatically set.

**Example proposal Bike panel**.
![Unified Bike arrival announcement](https://user-images.githubusercontent.com/54167201/230005095-70dfffc4-fb6b-424f-a4a0-f5dea42a5660.png)
- Other transport panels are similar.
- Each by prefered individual setting.

**Reply by devemux86** [reply]
@0709wiwiwi thanks for the suggestions!

Having all settings together indeed seems nice, but it becomes more complicated. :slightly_smiling_face: 
We currently separate the settings that have different options by navigation profile.

> Navigation instructions off ? = False
As there remain 2 announcements:
The Last announcement AND the Arrival announcement.

"Last announcement" and "Arrival announcement" are essential in voice guidance.
Do we need on / off options for them too?

Alternatively voice guidance can be turned off completely with the global switch.

**Reply by devemux86** [reply]
@0709wiwiwi thanks for the test!

To reproduce the test and check what happens, can you also post:
- App settings .pref file
- Route file

**Reply by 0709wiwiwi** [reply]
Files in zip.
+ contains the gpx file for the Lockito replay (fake gps) Android app.  (Speed set = 18kmh)
[Emux_Leen.zip](https://github.com/devemux86/cruiser/files/11157712/Emux_Leen.zip)

> Do we need options for them too?
I do set both of them activ in the example.   Makes it clear that bothe will be announced.

**Reply by devemux86** [reply]
@0709wiwiwi thanks for the samples, they make developer's life easier! :slightly_smiling_face: 

This is an interesting example of 2 consecutive turns close to each other: 19m.
You have enabled the "Settings | Voice guidance | Advanced | Current speed".

As explained [above](https://github.com/devemux86/cruiser/discussions/108#discussioncomment-5477196), this option allows the GPS speed to influence the voice guidance:
_speak the voice announcements some seconds before the position has been reached_.

In this particular example with the very close turns, the announcement of the 2nd turn is made before passing the 1st turn,
because of the time difference (caused by the speed influence). So we need an adjustment for such cases with this option.

Can you try again with disabled the "Settings | Voice guidance | Advanced | Current speed"?

Also with tests, turn off "Settings | Navigation | Advanced | Snap to route" to see the actual moving position.
(snapped positions are not actual positions, but projections along the route)

**Reply by 0709wiwiwi** [reply]
The same observation.  (18kmh is of course not a particularly fast speed)
By the way the Lockito track design is by the google map routing.  
There should so always be a very slight difference with the osm mapping..
If 100% exact the same path than someone copied Google into osm ;-) 
Google has a few 'hidden' non existing streets built into their map ;-) 
So a check for false Google copies is quickly observed ;-)

**Reply by devemux86** [reply]
[Cruiser 3.0.17](https://play.google.com/store/apps/details?id=gr.talent.cruiser) has an improved implementation of "Last announcement" to avoid such situations.

More improvements are planned, if you have specific examples please post the route and app settings .pref file.

**Reply by 0709wiwiwi** [reply]
Overall (both turns timing) is not as good as in the previous version.
- Last announcement entry = 60m.
In the previous version, the first turn was announced very correctly just before reaching the 60m distance 
By V.3.017 = The first announcement is very delayed even just **after** the 30m distance indication. 
**This especially is so not good.** _The second nearby turn is now timed almost correctly._
If optimisation is too difficult advice to bring back the previous implementation. 
_I can live with this the reported minor inconvenience anyway. 
 I (beta tester) report what I do observe, it's not a complaint_

**Reply by devemux86** [reply]
@0709wiwiwi thanks for the tests!

> If optimisation is too difficult advice to bring back the previous implementation.

Agreed, previous implementation was active and tested all these years.
We can return to it until more successful improvements can be made. :slightly_smiling_face:

**Reply by DolomitesTom** [reply]
The topic is not really important for me. Everything that is not disturbing is ok. What is more important to me is the average speed 🙄🙂.

**Reply by 0709wiwiwi** [reply]
To be clear:  The previous version implementation by v3.0.16 was fine.  (imo)   (Just that very last improvement is not right)

**Reply by devemux86** [reply]
> The previous version implementation by v3.0.16

Thanks, that's what I meant too. :slightly_smiling_face:

**Reply by devemux86** [reply]
> If optimisation is too difficult advice to bring back the previous implementation.

Available again in [Cruiser 3.0.18](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by DolomitesTom** [comment]
In this version the information "Route is being calculated" is no longer suppressed in voice navigation when you leave the planned route. Since using the intended diversions or alternative route does not work very well, this is annoying in the long run. In my opinion, this "calculate route" should not be pronounced at all.

**Reply by WolfgangZ42** [reply]
I think you can suppress it:
<img width="200" src="https://user-images.githubusercontent.com/121806238/230705123-5b9cc560-dc6b-45bc-a381-974ea2cd7b8c.png"/> 
Up to now it wasn't annoying me at all.

**Reply by devemux86** [reply]
Indeed as @WolfgangZ42 mentioned, the off route voice messages are controlled by the "Off route warning" setting:
- "Route missed" when rerouting is off
- "Calculate route" when rerouting is on

**Reply by DolomitesTom** [reply]
Ah okay thanks
